### ISSUE TITLE
Terminal pane starts up before Python extension and virtual environment isn't activated

### ISSUE THREAD

**Issue by pfmoore** [microsoft/vscode#123295]
- VS Code Version: 1.56.0 (user setup)
- OS Version: Windows 10

From Help/About:
```
Version: 1.56.0 (user setup)
Commit: cfa2e218100323074ac1948c885448fdf4de2a7f
Date: 2021-05-04T22:09:06.405Z
Electron: 12.0.4
Chrome: 89.0.4389.114
Node.js: 14.16.0
V8: 8.9.255.24-electron.0
OS: Windows_NT x64 10.0.19041
```
Steps to Reproduce:

1. Start VS Code on a directory containing a Python project, with a virtual environment `.venv` which is set as the default Python interpreter.
2. Start the integrated terminal, which will run with `.venv` activated as expected.
3. Close VS Code
4. Reopen VS Code
5. The editor will open with the terminal pane open, as it was previously closed with the terminal still open. Note that the Python extension is still starting at this point (you may not see this if your machine is particularly fast, but it is consistent on my PC).
6. The virtual environment is *not* activated, because the Python extension hasn't initialised.

I checked my settings, and I see

```
"terminal.integrated.enablePersistentSessions": true
"terminal.integrated.environmentChangesRelaunch": true
```

I did not set either of these myself, so I assume they are default values. They seem to suggest that what I should expect is for my terminal session to get restarted once the Python extension has loaded.

Does this issue occur when all extensions are disabled?: No

No, it's specifically related to how extensions initialise.

> 🪓 If you answered No above, use 'Help: Start Extension Bisect' from Command Palette to try to identify the cause.

I don't see how this would help, but if I'm missing something I'll be happy to do this with a little clarification as to what I should do.

> 📣 Issues caused by an extension need to be reported directly to the extension publisher. The 'Help > Report Issue' dialog can assist with this.

I don't believe this is specific to the Python extension, but I don't have another extension that changes the terminal environment, so I don't have any way to confirm this.

**Comment by meganrogge** [comment]
Can you share your settings?

Was this happening in previous versions?

**Comment by meganrogge** [comment]
Ctrl shift p -> Open Settings (JSON) or click the open settings button in the upper right hand corner of the settings editor.

**Comment by pfmoore** [comment]
Thanks! I should have thought of Ctrl-Shift-P...

```
{
    "workbench.iconTheme": "vscode-icons",
    "extensions.ignoreRecommendations": false,
    "vsicons.dontShowNewVersionMessage": true,
    "editor.fontFamily": "Cascadia Code, Consolas, 'Courier New', monospace",
    "powershell.integratedConsole.showOnStartup": false,
    "terminal.external.windowsExec": "pwsh",
    "python.languageServer": "Pylance",
    "python.showStartPage": false,
    "workbench.startupEditor": "newUntitledFile",
    "terminal.integrated.shell.windows": "pwsh.exe",
    "python.pythonPath": "C:\\Users\\Gustav\\AppData\\Local\\Programs\\Python\\Python39\\python.exe",
    "workbench.editorAssociations": [
        {
            "viewType": "jupyter.notebook.ipynb",
            "filenamePattern": "*.ipynb"
        }
    ],
}
```

**Comment by FitRiver** [comment]
@meganrogge I use different settings, but I have the same issue.

Previously, I used the following workspace settings to auto-activate the environment in the terminal:

    "terminal.integrated.shellArgs.linux": [
        "/path/to/activate.sh",
    ],

It activated the environment right after starting up.

These settings were deprecated in the latest version:

https://code.visualstudio.com/updates/v1_56#_terminal

I tried to use the new settings:

    "terminal.integrated.profiles.linux": {
        "bash": {
            "path": "bash",
            "args": [
                "/path/to/activate.sh",
            ],
        },
    },

Unfortunately, it only works for New Terminals or when I have a different tab open after the start and switch to Terminal after VS Code is fully loaded.

**Comment by meganrogge** [comment]
This is fixed now and in 1.56.2

**Comment by FitRiver** [comment]
@meganrogge Which settings are supposed to work in 1.56.2?

Now I tried that the old settings work on Linux:

    "terminal.integrated.shellArgs.linux": [
        "/path/to/activate.sh",
    ],

But it's marked as deprecated.

On the other hand the new syntax doesn't work:

    "terminal.integrated.profiles.linux": {
        "bash": {
            "path": "bash",
            "args": [
                "/path/to/activate.sh",
            ],
        },
    },

On Windows, neither the old syntax nor the new one worked (with 1.56.2).

Old:

    "terminal.integrated.shellArgs.windows": [
        "powershell -noexit -command \"& {cd $env:VSCODE_CWD; . C:/.../Scripts/Activate.ps1}\""
    ],

New:

    "terminal.integrated.profiles.windows": {
        "PowerShell": {
            "source": "PowerShell",
            "args": [
                "-noexit",
                "-command",
                "\"& {cd $env:VSCODE_CWD; . C:/.../Scripts/Activate.ps1}\"",
            ]
        }
    },

By not working I mean it's not activated on startup. It's activated when the new Terminal is launched (or the VS Code is started with hidden Terminal and the tab is switched to it).

**Comment by pfmoore** [comment]
@FitRiver I'm assuming that what is fixed is the behaviour I reported (the terminal doesn't wait to open until the Python extension is ready, so that it gets the activated virtual environment). What you're describing is different, and should probably be reported as a separate issue (unless there's some link that I'm not seeing - I'm just a user, @meganrogge can probably advise better than I can).

**Comment by pfmoore** [comment]
Hmm, @meganrogge I just tested the behaviour again - I now have 1.56.2 - and it's still happening.

Could you re-open this issue, as it looks like it hasn't been fixed?

**Comment by FitRiver** [comment]
> @FitRiver I'm assuming that what is fixed is the behaviour I reported (the terminal doesn't wait to open until the Python extension is ready, so that it gets the activated virtual environment).

I have been waiting on any response to your comment, but you are probably right it's a different issue. I'll open a new ticket.

**Comment by Tyriar** [comment]
Two things you might be hitting are that if shell or shellArgs is set, they will take precedence over the defaultProfile setting. Additionally the initial profile restored will not restore defaultProfile https://github.com/microsoft/vscode/issues/123188

The latter sounds like your issue @FitRiver. Note also that your args are a different format, in the first you're running powershell _within_ your shell with the args as a single string, the second is separate args within a single shell. This would be the equivalent, if your shell was powershell before:

```json
"terminal.integrated.profiles.windows": {
    "PowerShell": {
        "source": "PowerShell",
        "args": [
            "powershell -noexit -command \"& {cd $env:VSCODE_CWD; . C:/.../Scripts/Activate.ps1}\""
        ]
    }
},
```

**Comment by pfmoore** [comment]
@Tyriar I don't have those settings, as you can see from my settings posted above.

Can this issue please be reopened as it definitely hasn't been fixed for me? If not, I can open a new issue pointing back to this one - but I'd rather not have to do that as it seems unproductive compared to just re-opening this.

**Comment by Tyriar** [comment]
@pfmoore please try on the [Insiders build ](https://code.visualstudio.com/insiders) first, a lot of changes have gone into profiles since v1.56.2

**Comment by pfmoore** [comment]
Nope, still an issue with insiders build.

1. Install Insiders build
2. Install Python extension (auto-installs pylance and jupyter). I got asked if I wanted to use the insiders build of the extensions, I said yes.
3. Open a folder that contains a `.venv` directory with a virtual environment, and a `.vscode` directory with a `settings.json` containing just the line `"python.pythonPath": ".venv\\Scripts\\python.exe"`.
4. Open a .py file in that folder.
5. Open the terminal with `Ctrl-'`. Terminal starts and runs the venv activate script. Terminal is pwsh (Powershell core). I assume it's picking that up from my settings.
6. Close VSCode, *without closing the terminal*.
7. Restart VSCode. It starts with the previous folder open, and the terminal pane open as that's how it was when last closed. The terminal opens, but it starts opening *before* the Python extension has finished initialising (the "Python 3.9.2 64-bit ('.venv')" note in the bottom status bar hasn't appeared yet). When the terminal finishes starting up, it does *not* have the venv activated.

Why do you think it's related to profiles? It seems to me that the issue here is that when starting, if the terminal was remembered as having been open when last closed, VSCode restarts the terminal without waiting for extensions to initialise so that they can add their "do this when the terminal opens" settings. Is that profile-related?

**Comment by Tyriar** [comment]
> Restart VSCode. It starts with the previous folder open, and the terminal pane open as that's how it was when last closed. The terminal opens, **but it starts opening before the Python extension has finished initialising** (the "Python 3.9.2 64-bit ('.venv')" note in the bottom status bar hasn't appeared yet). When the terminal finishes starting up, it does not have the venv activated.

This is by design. I believe the Python team is looking into the better way of contributing to the environment via the `context.environmentVariableCollection` API. Not sure if they had an issue but created https://github.com/microsoft/vscode-python/issues/16318

**Comment by pfmoore** [comment]
I'm guessing that when you say "by design" you mean that a previously opened terminal is reopened without waiting for extensions because it was a deliberate choice not to wait. I can see that, but it has the unfortunate consequence that extensions that the way the Python extension activates the selected venv doesn't work.

Given that this means it's regrettably far too easy to end up working in a terminal with the wrong Python environment active, is there a way to work around this? I'm specifically thinking of a setting that lets me say "always start VS Code with the terminal window closed, even if it was open when the program last shut down". That would be for me far better than the current behaviour.

**Comment by Tyriar** [comment]
> I'm specifically thinking of a setting that lets me say "always start VS Code with the terminal window closed, even if it was open when the program last shut down". That would be for me far better than the current behaviour.

The issue I linked to should hopefully result in the best of both worlds, also it's possible the Python extension could use the current mechanism to activate within restored terminals since they can access them. Both of these would be jobs for the Python terminal though.

The setting idea has come up before but we'd prefer not to do that, regardless it wouldn't fix the case where you open a terminal quickly after launching (before extension activation).
